metaflow/metaflow_config.py

```diff
@@ -7,363 +6,278 @@
+from metaflow.metaflow_config_funcs import from_conf, get_validate_choice_fn
 
 # Disable multithreading security on MacOS
 if sys.platform == "darwin":
     os.environ["OBJC_DISABLE_INITIALIZE_FORK_SAFETY"] = "YES"
 
 
-def init_config():
-    # Read configuration from $METAFLOW_HOME/config_<profile>.json.
-    home = os.environ.get("METAFLOW_HOME", "~/.metaflowconfig")
-    profile = os.environ.get("METAFLOW_PROFILE")
-    path_to_config = os.path.join(home, "config.json")
-    if profile:
-        path_to_config = os.path.join(home, "config_%s.json" % profile)
-    path_to_config = os.path.expanduser(path_to_config)
-    config = {}
-    if os.path.exists(path_to_config):
-        with open(path_to_config) as f:
-            return json.load(f)
-    elif profile:
-        raise MetaflowException(
-            "Unable to locate METAFLOW_PROFILE '%s' in '%s')" % (profile, home)
-        )
-    return config
-
-
-# Initialize defaults required to setup environment variables.
-METAFLOW_CONFIG = init_config()
-
-
-def from_conf(name, default=None, validate_fn=None):
-    """
-    First try to pull value from environment, then from metaflow config JSON.
-
-    Prior to a value being returned, we will validate using validate_fn (if provided).
-    Only non-None values are validated.
-
-    validate_fn should accept (name, value).
-    If the value validates, return None, else raise an MetaflowException.
-    """
-    value_from_env = os.environ.get(name, None)
-    value_from_config = METAFLOW_CONFIG.get(name, default)
-    if value_from_env is not None:
-        value = value_from_env
-    else:
-        value = value_from_config
-    if validate_fn and value is not None:
-        validate_fn(name, value)
-    return value
-
-
-def _get_validate_choice_fn(choices):
-    """Returns a validate_fn for use with from_conf().
-    The validate_fn will check a value against a list of allowed choices.
-    """
-
-    def _validate_choice(name, value):
-        if value not in choices:
-            raise MetaflowException(
-                "%s must be set to one of %s. Got '%s'." % (name, choices, value)
-            )
-
-    return _validate_choice
-
-
 ###
 # Default configuration
 ###
-DEFAULT_DATASTORE = from_conf("METAFLOW_DEFAULT_DATASTORE", "local")
-DEFAULT_ENVIRONMENT = from_conf("METAFLOW_DEFAULT_ENVIRONMENT", "local")
-DEFAULT_EVENT_LOGGER = from_conf("METAFLOW_DEFAULT_EVENT_LOGGER", "nullSidecarLogger")
-DEFAULT_METADATA = from_conf("METAFLOW_DEFAULT_METADATA", "local")
-DEFAULT_MONITOR = from_conf("METAFLOW_DEFAULT_MONITOR", "nullSidecarMonitor")
-DEFAULT_PACKAGE_SUFFIXES = from_conf("METAFLOW_DEFAULT_PACKAGE_SUFFIXES", ".py,.R,.RDS")
-DEFAULT_AWS_CLIENT_PROVIDER = from_conf("METAFLOW_DEFAULT_AWS_CLIENT_PROVIDER", "boto3")
+
+# Propagate defaults to be able to read what the user sets in configuration files/envs;
+# this will *not* propagate things set on the command line
+DEFAULT_DATASTORE = from_conf("DEFAULT_DATASTORE", "local")
```

Comment:
Can we add a top level comment to this file that calls out that the actual env vars are `METAFLOW_*`? Alternatively, we can have from_conf("METAFLOW_DEFAULT_ENVIRONMENT") etc. so that it is easier for metaflow users to find out what exact env vars are used with Metaflow.

Reply:
I will add a comment.